File: ios/Classes/Log.swift
//
//  Log.swift
//  flutter_label_printer
//
//  Created by Peter Wong (Engineering) on 8/2/2023.
//

import Foundation
import os.log

public class Log: NSObject {
    enum LogLevel: Int, CaseIterable {
        case verbose, debug, info, warn, error, assert
    }
    
    let log = OSLog(subsystem: Bundle.main.bundleIdentifier!, category: "FlutterLabelPrinter")

    var logLevel = LogLevel.info
    
    public func setLogLevel(level: Int) {
        let maxLevel = LogLevel.allCases.count
        let aLevel = max(0, min(level, maxLevel))
        logLevel = LogLevel(rawValue: aLevel)!
    }
    
    public func v(msg: StaticString) {
        if (logLevel.rawValue <= LogLevel.verbose.rawValue) {
            os_log(msg, type: .debug)
        }
    }
    
    public func d(msg: StaticString) {
        if (logLevel.rawValue <= LogLevel.debug.rawValue) {
            os_log(msg, type: .debug)
        }
    }
    
    public func i(msg: StaticString) {
        if (logLevel.rawValue <= LogLevel.info.rawValue) {
            os_log(msg, type: .info)
        }
    }
    
    public func w(msg: StaticString) {
        if (logLevel.rawValue <= LogLevel.warn.rawValue) {
            os_log(msg, type: .default)
        }
    }
    
    public func e(msg: StaticString) {
        if (logLevel.rawValue <= LogLevel.error.rawValue) {
            os_log(msg, type: .error)
        }
    }
    
    public func wtf(msg: StaticString) {
        if (logLevel.rawValue <= LogLevel.assert.rawValue) {
            os_log(msg, type: .fault)
        }
    }
}

File: android/src/main/kotlin/hk/gogovan/flutter_label_printer/util/Log.kt
package hk.gogovan.flutter_label_printer.util

class Log {
    companion object {
        const val LOG_TAG = "FlutterLabelPrinter"
    }

    enum class LogLevel {
        VERBOSE, DEBUG, INFO, WARN, ERROR, ASSERT
    }

    var logLevel: LogLevel = LogLevel.INFO

    fun setLogLevel(level: Int) {
        val maxLevel = LogLevel.values().size
        val aLevel = if (level < 0) 0 else if (level > maxLevel) maxLevel else level
        logLevel = LogLevel.values()[aLevel]
    }

    fun v(msg: String) {
        if (logLevel <= LogLevel.VERBOSE) {
            android.util.Log.v(LOG_TAG, msg)
        }
    }

    fun d(msg: String) {
        if (logLevel <= LogLevel.DEBUG) {
            android.util.Log.d(LOG_TAG, msg)
        }
    }

    fun i(msg: String) {
        if (logLevel <= LogLevel.INFO) {
            android.util.Log.i(LOG_TAG, msg)
        }
    }

    fun w(msg: String) {
        if (logLevel <= LogLevel.WARN) {
            android.util.Log.w(LOG_TAG, msg)
        }
    }

    fun e(msg: String) {
        if (logLevel <= LogLevel.ERROR) {
            android.util.Log.e(LOG_TAG, msg)
        }
    }

    fun wtf(msg: String) {
        if (logLevel <= LogLevel.ASSERT) {
            android.util.Log.wtf(LOG_TAG, msg)
        }
    }

}